libgadget/lenstools.c

```diff
@@ -0,0 +1,336 @@
+#include <mpi.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <math.h>
+#include <fftw3.h> 
+#include <string.h>
+
+#ifdef USE_CFITSIO
+#include "fitsio.h"
+#endif
+
+#include "lenstools.h"
+#include "partmanager.h"
+#include "cosmology.h"
+#include "physconst.h"
+#include "utils.h"
+
+void linspace(double start, double stop, int num, double *result) {
+    double step = (stop - start) / (num - 1);
+    for (int i = 0; i < num; i++) {
+        result[i] = start + i * step;
+    }
+}
+
+// Function to allocate a 1D array to be used as a 3D array, and initialize elements to zero
+double *allocate_3d_array_as_1d(int Nx, int Ny, int Nz) {
+    double *array = mymalloc("3d", Nx * Ny * Nz * sizeof(double));
+    // Initialize elements with memset
+    memset(array, 0, Nx * Ny * Nz * sizeof(double));
+    return array;
+}
+
+// Function to allocate a 1D array to be used as a 2D array, and initialize elements to zero
+double *allocate_2d_array_as_1d(int Nx, int Ny) {
+    double *array = mymalloc("2d", Nx * Ny * sizeof(double));
+    // Initialize elements with memset
+    memset(array, 0, Nx * Ny * sizeof(double));
+
+    return array;
+}
+
+typedef struct {
+    int nx, ny, nz;
+} GridDimensions;
+
+// Function to determine the bin index for a given value
+int find_bin(double value, double *bins, int resolution, const double L) { // L is the box size
+    // float index
+    float iflt = (value - bins[0]) / (bins[resolution] - bins[0]) * resolution;
+
+    // round down to the nearest integer
+    int index = (int)floorf(iflt);
+
+    // check if the value is within the range
+    if (index >= 0 && index < resolution) {
+        return index;
+    }
+    else {
+        return -1;
+    }
+    endrun(0, "Error in find_bin\n");
```

Comment:
Don't need the endrun because it either returns -1 or index so endrun will never be reached.

Reply:
Right, let me remove it.